chalk-slg/src/lib.rs

```diff
@@ -0,0 +1,322 @@
+pub struct SimplifiedAnswer<C: Context> {
+    /// A fully instantiated version of the goal for which the query
+    /// is true (including region constraints).
+    pub subst: C::CanonicalConstrainedSubst,
+
+    /// If this flag is set, then the answer could be neither proven
+    /// nor disproven. In general, the existence of a non-empty set of
+    /// delayed literals simply means the answer's status is UNKNOWN,
+    /// either because the size of the answer exceeded `max_size` or
+    /// because of a negative loop (e.g., `P :- not { P }`).
+    pub ambiguous: bool,
+}
+
+#[derive(Clone, Debug)]
+enum DelayedLiteralSets<C: Context> {
+    /// Corresponds to a single, empty set.
+    None,
+
+    /// Some (non-zero) number of non-empty sets.
+    Some(HashSet<DelayedLiteralSet<C>>),
+}
+
+/// A set of delayed literals. The vector in this struct must
+/// be sorted, ensuring that we don't have to worry about permutations.
+///
+/// (One might expect delayed literals to always be ground, since
+/// non-ground negative literals result in flounded
+/// executions. However, due to the approximations introduced via RR
+/// to ensure termination, it *is* in fact possible for delayed goals
+/// to contain free variables. For example, what could happen is that
+/// we get back an approximated answer with `Goal::CannotProve` as a
+/// delayed literal, which in turn forces its subgoal to be delayed,
+/// and so forth. Therefore, we store canonicalized goals.)
+#[derive(Clone, Debug, Default, PartialEq, Eq, Hash)]
+struct DelayedLiteralSet<C: Context> {
+    delayed_literals: Vec<DelayedLiteral<C>>,
+}
+
+#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub enum DelayedLiteral<C: Context> { // FIXME: pub b/c fold
+    /// Something which can never be proven nor disproven. Inserted
+    /// when truncation triggers; doesn't arise normally.
+    CannotProve(()),
+
+    /// We are blocked on a negative literal `~G`, where `G` is the
+    /// goal of the given table. Because negative goals must always be
+    /// ground, we don't need any other information.
+    Negative(TableIndex),
+
+    /// We are blocked on a positive literal `Li`; we found a
+    /// **conditional** answer (the `CanonicalConstrainedSubst`) within the
+    /// given table, but we have to come back later and see whether
+    /// that answer turns out to be true.
+    Positive(TableIndex, C::CanonicalConstrainedSubst),
+}
+
+/// Either `A` or `~A`, where `A` is a `Env |- Goal`.
+#[derive(Clone, Debug)]
+pub enum Literal<C: Context> { // FIXME: pub b/c fold
+    Positive(C::GoalInEnvironment),
+    Negative(C::GoalInEnvironment),
+}
+
+impl<C: Context> PartialEq for Literal<C> {
+    fn eq(&self, other: &Literal<C>) -> bool {
+        match (self, other) {
+            (Literal::Positive(goal1), Literal::Positive(goal2))
+            | (Literal::Negative(goal1), Literal::Negative(goal2)) => goal1 == goal2,
+
+            _ => false,
+        }
+    }
+}
```

Comment:
Is there a reason for not `#[derive]`-ing these impls?

Reply:
Gah. I'm torn. On the one hand, they are not *needed*, and it requires deriving a bunch of ord and crap elsewhere that are also not *needed*.

On the *other* hand, writing the impls by hand is a real pain. After having tried that for long enough, I'm maybe inclined to just add the derives.

Reply:
actually I was being silly. Fixed.